### ISSUE TITLE
(Possible) bug in strategy 6

### ISSUE THREAD

**Issue by maxecoulter** [eddelbuettel/rcppde#17]
Hi,
My understanding is that rcppde was developed from DEoptim a while back (https://cran.irsn.fr/web/packages/RcppDE/vignettes/RcppDE.pdf).
I believe I have found a bug in the switch statement (specifically for case 6, lines 172 - 179 in devol.cpp). 

The code is currently:
`

      case 6: {                

                int i_pbest = sortIndex[static_cast<int>(::unif_rand() * p_NP)]; // select from [0, 1, 2, ..., (pNP-1)] 

                int j = static_cast<int>(::unif_rand() * i_D);  // random parameter 

                do {                            // add fluctuation to random target 

                    t_tmpP[j] = ta_oldP.at(j,i) + rand_weight * (ta_oldP.at(j,i_pbest) - ta_oldP.at(j,i)) + 
                        rand_weight * (ta_oldP.at(j,ia_urn2[1]) - ta_oldP.at(j,ia_urn2[2]));

                    j = (j + 1) % i_D;

                } while ((::unif_rand() < rand_cross) && (++k < i_D));

                break;

            }`
However, if you look at the C code in the DEoptim version (https://github.com/ArdiaD/DEoptim/blob/master/src/de4_0.c, lines 372 - 379) the statement ` i_pbest = sortIndex[(int)(unif_rand() * p_NP)];` occurs within the while loop.

`

        case 6: { /*---DE/current-to-p-best/1 (JADE)----------------------*/

            /* select from [0, 1, 2, ..., (pNP-1)] */

            i_pbest = sortIndex[(int)(unif_rand() * p_NP)];

            nt_tmpP[i+i_NP*j] = ngta_oldP[i+i_NP*j] +
              d_weight * (ngta_oldP[i_pbest+i_NP*j] - ngta_oldP[i+i_NP*j]) +
              d_weight * (ngta_oldP[i_r1+i_NP*j]    - ngta_oldP[i_r2+i_NP*j]);

            break;

          }`
 So in the Rcppde version, one best solution is chosen for the whole of the loop, while in the C version a different best solution is chosen for every value of j.
 
 If you do this instead:
 
 `

      case 6: {                           // ---DE/current-to-p-best/1 (JADE)-----------------------------------

                i_pbest;

                int j = static_cast<int>(::unif_rand() * i_D);  // random parameter 

                do {                            // add fluctuation to random target 

                     i_pbest = sortIndex[static_cast<int>(::unif_rand() * p_NP)]; // select from [0, 1, 2, ..., (pNP-1)] 

                    t_tmpP[j] = ta_oldP.at(j,i) + rand_weight * (ta_oldP.at(j,i_pbest) - ta_oldP.at(j,i)) + 
                        rand_weight * (ta_oldP.at(j,ia_urn2[1]) - ta_oldP.at(j,ia_urn2[2]));

                    j = (j + 1) % i_D;

                } while ((::unif_rand() < rand_cross) && (++k < i_D));

                break;

            }`
This solves the problem.

I was finding that the original Rcppde code with stategy 6 was converging too early, well short of the actual solution. The above fix solved it so it has similar performance with DEoptim.

However, I also noticed a few other differences between the DEoptim and Rcppde methods. I presume the C code of DEoptim has been added to since rcppde was made (last fix was very recently). E.g (at line 434 in https://github.com/ArdiaD/DEoptim/blob/master/src/de4_0.c):

`

     if (d_c > 0) { /* calculate new goodCR and goodF */

          goodCR += d_cross / ++i_goodNP;

          goodF += d_weight;

          goodF2 += pow(d_weight,2.0);

        }`
        
 There are others as well. But to be honest I don't exactly know the effect of these new changes, or whether they have intentionally not been included. I have good performance with these extra bits. I have NOT tested the switch fix without these extra additions, though I can do if you wish. I have tested it the other way around, with the extra additions without the switch statement fix, it doesn't make much difference compared to the original version. If you would like my version of devol.cpp I can get this to you (I can't seem to attach it here).
 
 So in summary:
 1) The difference between DEoptim and Rcppde in the switch statement looks like a bug in the RcppDE code, it has a big performance implication in strategy 6.
 2) There are other minor differences, but these look like addins to DEoptim. I don't fully know the implications of adding these in/ leaving them out. I have added them in anyway.
 
 Let me know how you would like to proceed.
 
 Best regards,
 
 Max
 
 Dr Max Coulter

Bioinformatician/Geneticist

XELECT LTD
HORIZON HOUSE
ABBEY WALK
ST ANDREWS // FIFE
KY16 9LB
SCOTLAND // UK

W www.xelect.co.uk

T +44 (0) 7598 930557

E [max.coulter@xelect.co.uk](mailto:max.coulter@xelect.co.uk)

**Comment by eddelbuettel** [comment]
Thanks for taking a detailed look, and yes, as you can see from the commit log, the code here may now trail `DEoptim`.  I think my preference would be to go issue by issue, so maybe we tackle what you identified above as the change in strategy first (and in isolation) in a pull request and move from there.  Would that work for you?

**Comment by maxecoulter** [comment]
Sounds good to me. So I'll create a pull request for the switch statement issue .

**Comment by eddelbuettel** [comment]
Sounds good.   One wish, if I may, is to please also update ChangeLog with a short note. Emacs makes it semi-automatic, in other editors is easy enough too,

**Comment by maxecoulter** [comment]
Ah I did add to to the changelog, but was having issues with github and vs-code - I'll add that in, sorry

**Comment by eddelbuettel** [comment]
Seems to have come through fine as two commits.  Still marked as draft though, and I added one more question.

**Comment by maxecoulter** [comment]
Hi Dirk,
Is there anything else I need to do regarding the pull request?
I ran this new version with my data, and I find it still performs poorly. Its only when you add in those extra bits present in DEoptim (goodCR/goodF etc) that it gets to where it needs to be.
Unfortunately this is all a bit cryptic, I can't share my data/ objective functions with you as it belongs to my company. And I'm guessing strategy 6 may well perform fine with your tests/ objective functions. So you may well want to leave it here, but just letting you know it may be worth exploring.
Max

**Comment by eddelbuettel** [comment]
My apologies.  This just fell to the side when I got distracted.  Will merge this now, and I suggest we then move the next suggested change relative to where DEoptim is now.

**Comment by eddelbuettel** [comment]
Done, and I also increased the micro version to .2 to set this apart.  Thanks again.

**Comment by maxecoulter** [comment]
No worries! I probably won't have time today, but would you like me to push up a version with the further DEoptim additions for you to check? Or you can do it if you want to take things from here...I don't mind either way.

**Comment by eddelbuettel** [comment]
I like what we did with the focussed and clear PR.  If you can isolate different changesets (as you see fit, ideally somewhere between 'one large waterfall' (too coarse) and 'one line each' (too granular) we can knock this out calmly and, once done, release up to CRAN.  I really appreciate that you are "in the weeds" for this one.  Not urgent for me either, but we should do this. (And we may want a new issue ticket per actual distinct issue, may be cleaner than extending this one.  But either way works, really.)

**Comment by maxecoulter** [comment]
Sounds good - I'll have a look at it next week.

**Comment by eddelbuettel** [comment]
Sorry for the delays, and thanks again for your two very fine PRs.  I think this should now be ready for a quick update at CRAN, or is there anything else you could or would think of?
